core/test/gtest/environments.hpp

```diff
@@ -0,0 +1,220 @@
+        auto rs_count = rs_count_env ? std::stoi(rs_count_env) : 0;
+        if (rs_count == 0) {
+            std::cerr << "Running without CTest ctest_resource configuration"
+                      << std::endl;
+            return;
+        }
+        if (rs_count != size) {
+            GKO_INVALID_STATE("Invalid resource group count: " +
+                              std::to_string(rs_count));
+        }
+
+        // parse CTest ctest_resource group descriptions
+        if (rank == 0) {
+            std::cerr << "Running with CTest ctest_resource configuration:"
+                      << std::endl;
+        }
+        // OpenMP CPU threads
+        if (auto rs_omp_env = get_ctest_group("cpu", rank)) {
+            auto resource = parse_ctest_resources(rs_omp_env);
+            omp_threads = resource.slots;
+            if (rank == 0) {
+                std::cerr << omp_threads << " CPU threads" << std::endl;
+            }
+        }
+        // CUDA GPUs
+        if (auto rs_cuda_env = get_ctest_group("cudagpu", rank)) {
+            auto resource = parse_ctest_resources(rs_cuda_env);
+            cuda_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": CUDA device " << cuda_device_id
+                      << std::endl;
+        }
+        // HIP GPUs
+        if (auto rs_hip_env = get_ctest_group("hipgpu", rank)) {
+            auto resource = parse_ctest_resources(rs_hip_env);
+            hip_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": HIP device " << cuda_device_id
+                      << std::endl;
+        }
+        // SYCL GPUs (no other devices!)
+        if (auto rs_sycl_env = get_ctest_group("syclgpu", rank)) {
+            auto resource = parse_ctest_resources(rs_sycl_env);
+            sycl_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": SYCL device " << cuda_device_id
+                      << std::endl;
+        }
+    }
+
+    static int omp_threads;
+    static int cuda_device_id;
+    static int hip_device_id;
+    static int sycl_device_id;
+};
+
+
+#ifdef GKO_COMPILING_OMP
+
+class OmpEnvironment : public ::testing::Environment {
+public:
+    void SetUp() override
+    {
+        if (ResourceEnvironment::omp_threads > 0) {
+            omp_set_num_threads(ResourceEnvironment::omp_threads);
+        }
+    }
+};
+
+#else
+
+
+class OmpEnvironment : public ::testing::Environment {};
+
+#endif
+
+
+#ifdef GKO_COMPILING_CUDA
+
+class CudaEnvironment : public ::testing::Environment {
+public:
+    void TearDown() override
+    {
+        gko::kernels::cuda::reset_device(ResourceEnvironment::cuda_device_id);
+    }
+};
+
+#else
+
+class CudaEnvironment : public ::testing::Environment {};
+
+#endif
+
+
+#ifdef GKO_COMPILING_HIP
+
+class HipEnvironment : public ::testing::Environment {
+public:
+    void TearDown() override
+    {
+        gko::kernels::hip::reset_device(ResourceEnvironment::hip_device_id);
```

Comment:
The TearDown are executed from several ranks, right?
If a gpu is assigned to several rank, the reset_device will be called several times.
ResetDevice does not contain the counter now, so it leads the issue.

Reply:
I don't think this can happen. Each rank owns a single gpu exclusively.

Reply:
Yes, the MPI processes are completely independent, so device resets don't influence each other

Reply:
three ranks on the same GPU if test on amdci
but from https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gef69dd5c6d0206c2b8d099abac61f217, it only reset the associated stuff from the current process. It should be okay, right?

Reply:
that should not happen. amdci has only one cuda/hip gpu, so the resource can't be allocated. I think it will then skip the test, but I have to check that.

Reply:
Although the reset device concern is resolved, we should not skip the test if there's only one GPU in the machine. Multiple ranks should be able to use the same GPU device.

Reply:
I checked quickly, there seems to be little runtime difference in whether we oversubscribe GPUs or not. Since we run MPI tests only on multi-GPU systems, and this only applies to CI jobs, I think exclusive GPU access should be more reliable.

For now, I replaced the RESOURCE_PERCENTAGE approach by a single slot per GPU. If we explicitly want to oversubscribe GPUs, we can configure that on the CI runner side with the resources.json file.